Sources/PackageGraph/PackageGraph+Loading.swift

```diff
@@ -441,7 +444,13 @@ private func createResolvedPackages(
 
         let productDependencyMap: [String: ResolvedProductBuilder]
         if lookupByProductIDs {
-            productDependencyMap = productDependencies.spm_createDictionary { ($0.product.identity, $0) }
+            productDependencyMap = productDependencies.spm_createDictionary {
+                guard let packageName = packageBuilder.dependencyNamesForTargetDependencyResolutionOnly[$0.packageBuilder.package.identity] else {
+                    fatalError()
```

Comment:
Not sure what to do in this case, theoretically shouldn't happen. So we could either just error or fall back to best effort of assuming the package name is equal to the product name.

Reply:
Decided to throw an internal error here.